crates/red_knot_python_semantic/resources/mdtest/declaration/with.md

```diff
@@ -0,0 +1,122 @@
+# With declarations
+
+## Basic `with` statement
+
+The type of the target variable in a `with` statement is the return type from the context manager's `__enter__` method.
+
+```py
+class Target: ...
+
+class Manager:
+    def __enter__(self) -> Target:
+        return Target()
+
+    def __exit__(self, exc_type, exc_value, traceback): ...
+
+with Manager() as f:
+    reveal_type(f)  # revealed: Target
+```
+
+## Union context manager
+
+```py
+def coinflip() -> bool:
+    return True
+
+class Manager1:
+    def __enter__(self) -> str:
+        return "foo"
+
+    def __exit__(self, exc_type, exc_value, traceback): ...
+
+class Manager2:
+    def __enter__(self) -> int:
+        return 42
+
+    def __exit__(self, exc_type, exc_value, traceback): ...
+
+context_expr = Manager1() if coinflip() else Manager2()
+
+with context_expr as f:
+    reveal_type(f)  # revealed: str | int
+```
+
+## Context manager without an `__enter__` or `__exit__` method
+
+```py
+class Manager: ...
+
+# error: [invalid-context-manager] "Object of type Manager cannot be used with `with` because it doesn't implement `__enter__` and `__exit__`"
```

Comment:
The diagnostic messages here and below are excellent! Super clear and informative.